storage/innobase/trx/trx0rseg.cc

```diff
@@ -571,12 +601,48 @@ trx_rseg_array_init()
 				ut_ad(!trx_sys.rseg_array[rseg_id]);
 				trx_sys.rseg_array[rseg_id] = rseg;
 				trx_rseg_mem_restore(rseg, max_trx_id, &mtr);
+#ifdef WITH_WSREP
+				if (!wsrep_sys_xid.is_null() &&
+				    !wsrep_sys_xid.eq(&trx_sys.recovered_wsrep_xid)) {
+					wsrep_xid_in_rseg_found = true;
+					ut_ad(memcmp(wsrep_xid_uuid(&wsrep_sys_xid),
+						     wsrep_xid_uuid(&trx_sys.recovered_wsrep_xid),
+						     sizeof wsrep_uuid) ||
+					      (wsrep_xid_seqno(&wsrep_sys_xid) <=
+					       wsrep_xid_seqno(&trx_sys.recovered_wsrep_xid)));
+				}
+#endif
 			}
 		}
 
 		mtr.commit();
 	}
 
+#ifdef WITH_WSREP
+	if (!wsrep_sys_xid.is_null()) {
+		/* Upgrade from a version prior to 10.3.5,
+		where WSREP XID was stored in TRX_SYS page.
+		If no rollback segment has a WSREP XID set,
+		we must copy the XID found in TRX_SYS page
+		to rollback segments. */
+		mtr_t mtr;
+		mtr.start();
+
+		if (!wsrep_xid_in_rseg_found) {
+			trx_rseg_update_wsrep_checkpoint(&wsrep_sys_xid, &mtr);
+		}
+
+		/* Finally, clear WSREP XID in TRX_SYS page. */
+		const buf_block_t* sys = trx_sysf_get(&mtr);
+		mlog_write_ulint(sys->frame + TRX_SYS +
+				 TRX_SYS_WSREP_XID_INFO +
+				 + TRX_SYS_WSREP_XID_MAGIC_N_FLD,
+				 0, MLOG_4BYTES, &mtr);
```

Comment:
For constant folding to work, it might be better to use syntax like `&sys->frame[…]` or `… + sys->frame`. Not a big deal.
When merging this to 10.4, I might write a `MLOG_MEMSET` record for clearing more of the `TRX_SYS` page.